### PR TITLE
Docs and refactoring

### PR DESCRIPTION
**Pull request by samdark**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ✔️
| Fixed issues  | #121

Docs:

- [x] IterableFilterHandler/*
- [x] CountableDataInterface
- [x] DataReaderException
- [x] DataReaderReaderInterface
- [x] FilterableDataReaderInterface
- [x] FilterAssertHelper
- [x] FilterInterface
- [x] IterableDataReader
- [x] IterableFilterHandlerInterface
- [x] OffsetableDataInterface
- [x] ReadableDataInterface
- [x] Sort
- [x] SortableDataInterface
- [x] README

Additional things:

- [x] Raise minimum stability

### CODE REVIEW

File: src/Reader/IterableFilterHandler/EqualsEmpty.php [hunk 1]
```diff
@@ -2,28 +2,32 @@
...
 use InvalidArgumentException;
-use Yiisoft\Data\Reader\FilterDataValidationHelper;
+use Yiisoft\Data\Reader\FilterAssertHelper;
+use Yiisoft\Data\Reader\IterableFilterHandlerInterface;
 
 use function count;
 
-final class EqualsEmpty implements IterableHandlerInterface
+/**
+ * EqualsEmpty iterable filter handler checks that the item's field value is empty.
```
Comment (vjik) [inline src/Reader/IterableFilterHandler/EqualsEmpty.php:14]:
```suggestion
 * `EqualsEmpty` iterable filter handler checks that the item's field value is empty.
```

File: src/Reader/IterableFilterHandler/EqualsNull.php [hunk 2]
```diff
@@ -2,29 +2,33 @@
...
 use Yiisoft\Arrays\ArrayHelper;
-use Yiisoft\Data\Reader\FilterDataValidationHelper;
+use Yiisoft\Data\Reader\FilterAssertHelper;
+use Yiisoft\Data\Reader\IterableFilterHandlerInterface;
 
 use function count;
 
-final class EqualsNull implements IterableHandlerInterface
+/**
+ * EqualsNull iterable filter handler checks that the item's field value is null.
```
Comment (vjik) [inline src/Reader/IterableFilterHandler/EqualsNull.php:15]:
```suggestion
 * `EqualsNull` iterable filter handler checks that the item's field value is null.
```

File: src/Reader/IterableFilterHandler/GreaterThan.php [hunk 3]
```diff
@@ -2,10 +2,13 @@
 
 declare(strict_types=1);
 
-namespace Yiisoft\Data\Reader\Iterable\Handler;
+namespace Yiisoft\Data\Reader\IterableFilterHandler;
 
 use DateTimeInterface;
 
+/**
+ * GreaterThan iterable filter handler checks that the item's field value is greater than the given value.
```
Comment (vjik) [inline src/Reader/IterableFilterHandler/GreaterThan.php:10]:
```suggestion
 * `GreaterThan` iterable filter handler checks that the item's field value is greater than the given value.
```

File: src/Reader/IterableFilterHandler/GreaterThanOrEqual.php [hunk 4]
```diff
@@ -2,10 +2,14 @@
 
 declare(strict_types=1);
 
-namespace Yiisoft\Data\Reader\Iterable\Handler;
+namespace Yiisoft\Data\Reader\IterableFilterHandler;
 
 use DateTimeInterface;
 
+/**
+ * GreaterThanOrEqual iterable filter handler checks that the item's field value
```
Comment (vjik) [inline src/Reader/IterableFilterHandler/GreaterThanOrEqual.php:10]:
```suggestion
 * `GreaterThanOrEqual` iterable filter handler checks that the item's field value
```

File: src/Reader/IterableFilterHandler/In.php [hunk 5]
```diff
@@ -2,11 +2,14 @@
...
 declare(strict_types=1);
 
-namespace Yiisoft\Data\Reader\Iterable\Handler;
+namespace Yiisoft\Data\Reader\IterableFilterHandler;
 
 use function in_array;
 use function is_array;
 
+/**
+ * In iterable filter handler ensures that the field value matches one of the value provided.
```
Comment (vjik) [inline src/Reader/IterableFilterHandler/In.php:11]:
```suggestion
 * `In` iterable filter handler ensures that the field value matches one of the value provided.
```

File: src/Reader/IterableFilterHandler/Like.php [hunk 6]
```diff
@@ -2,11 +2,14 @@
...
 declare(strict_types=1);
 
-namespace Yiisoft\Data\Reader\Iterable\Handler;
+namespace Yiisoft\Data\Reader\IterableFilterHandler;
 
 use function is_string;
 use function stripos;
 
+/**
+ * Like iterable filter handler ensures that the field value is like-match to a given value.
```
Comment (vjik) [inline src/Reader/IterableFilterHandler/Like.php:11]:
```suggestion
 * `Like` iterable filter handler ensures that the field value is like-match to a given value.
```

File: src/Reader/IterableDataReader.php [hunk 7]
```diff
@@ -2,32 +2,26 @@
 
 declare(strict_types=1);
 
-namespace Yiisoft\Data\Reader\Iterable;
+namespace Yiisoft\Data\Reader;
```
Comment (vjik) [inline src/Reader/IterableDataReader.php:5]:
```suggestion
namespace Yiisoft\Data\Reader\Iterable;
```

Seems, better keep namespace and don't mix common data reader contracts and classes with implementation (iterable reader and specific filter handlers).

**Review by vjik — CHANGES_REQUESTED** [review]

File: src/Reader/IterableDataReader.php [hunk 8]
```diff
@@ -38,29 +32,38 @@
 use function uasort;
 
 /**
+ * Iterable data reader takes an iterable data as a source and can:
+ *
+ * - Limit items read
+ * - Skip N items from the beginning
+ * - Sort items
+ * - Form a filter criteria with {@see FilterInterface}
+ * - Post-filter items with {@see IterableFilterHandlerInterface}
```
[the remark this answers is not in the document]
Comment (samdark) [reply src/Reader/IterableDataReader.php]: It does post-filtering of its items.

File: README.md [hunk 9]
```diff
@@ -58,15 +65,20 @@ foreach ($result as $item) {
...
 
-Number of items returned can be limited:
+### Limiting number of items to read
+
+Number of items in an iterator can be limited:
 
 ```php
 $reader = (new MyDataReader(...))->withLimit(10);
+foreach ($reader->read() as $item) {
+
```
Comment (vjik) [inline README.md:75]:
```suggestion
    // ...
```

**Review by vjik — APPROVED** [review]

**Pull request merged**